lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -404,6 +422,37 @@ void SingleExitLoopTransformer::initialize() {
       llvm::for_each(inst.getResults(), saveEscaping);
     }
   }
+
+  // Do not eliminate undefs unless requested for.
+  if (!TFNoUndefsInSESE)  return result;
+
+  // Build the equivalence classes induced by argument passing.
+  llvm::EquivalenceClasses<SILValue> equivalentValues;
+  for (auto &bb : *currentFn) {
+    for (auto arg : bb.getArguments()) {
+      SmallVector<SILValue, 8> incomingValues;
+      arg->getIncomingValues(incomingValues);
+      for (SILValue incomingValue : incomingValues) {
+        equivalentValues.unionSets(arg, incomingValue);
+      }
+    }
+  }
+
+  // Replace undef with an equivalent value that is available at preheader.
+  for (auto &kv : result) {
+    const SILValue &escapingValue = kv.first;
+    // Find an equivalent value that dominates the header.
+    for (auto equivalentValue :
+         make_range(equivalentValues.findLeader(escapingValue),
```

Comment:
is it possible that we have an equiv class [v1, v2], and v1 is chosen as the leader, but v2 dominates preheader, not v1?

Reply:
Yes, that can happen. The leader for a class depends upon a bunch of factors:
  - order in which unification constraints are processed
  - the implementation of the union-find

Reply:
Then it would seem to lead to suboptimal transformation -- if we choose v2 in the above example, the undef can be removed. Do we want to iterate over the members (as in your old code), rather than only checking on the "leader" picked by the underlying llvm infra? (pls feel free to scope it out of this PR.)

Reply:
We are actually iterating over all the elements of the equivalence class.  For this example, we will chose v2 as we iterate over {v1, v2}. See the make_range(...) call. 

(It seems to me that the code is not readable. Any suggestions to improve it?)

Reply:
So we are iterating between `equivalentValues.findLeader(escapingValue)` and `equivalentValues.member_end()`. say the container has a list of elems `[v1, v2, v3]`, and `findLeader()` returns `v2`, does that then skip processing `v1`?

I wonder if we should call `member_begin()` instead of `findLeader()`.

Reply:
Ah, I see the source of confusion now. findLeader() returns an member iterator for the equivalence class:

Reply:
Got on the same page with Gogul through offline discussion.